File: web/src/util/cpf.ts
// Utilitário de CPF no frontend — espelha Servicos/Cpf.cs do backend.
// O CPF é a chave civil da pessoa; validar aqui dá feedback imediato
// (Fail Fast no cliente) antes do POST. O backend valida de novo: a LEI
// é do servidor; isto é só cortesia de UX.

/** Mantém só os dígitos (remove pontos, traço, espaços). */
export function normalizarCpf(valor: string): string {
  return valor.replace(/\D/g, '')
}

/** Formata para exibição: '00000000000' -> '000.000.000-00'. */
export function formatarCpf(valor: string): string {
  const d = normalizarCpf(valor)
  if (d.length !== 11) return valor
  return `${d.slice(0, 3)}.${d.slice(3, 6)}.${d.slice(6, 9)}-${d.slice(9)}`
}

/** Valida pelos dois dígitos verificadores (algoritmo da Receita). */
export function cpfValido(valor: string): boolean {
  const d = normalizarCpf(valor)
  if (d.length !== 11) return false
  if (d.split('').every((c) => c === d[0])) return false

  const digito = (parcial: string, pesoInicial: number): number => {
    let soma = 0
    for (let i = 0; i < parcial.length; i++) {
      soma += Number(parcial[i]) * (pesoInicial - i)
    }
    const resto = soma % 11
    return resto < 2 ? 0 : 11 - resto
  }

  const primeiro = digito(d.slice(0, 9), 10)
  const segundo = digito(d.slice(0, 10), 11)
  return primeiro === Number(d[9]) && segundo === Number(d[10])
}

File: api/SGI.Api/Servicos/Cpf.cs
namespace SGI.Api.Servicos;

/// <summary>
/// Utilitário de CPF. Com a refatoração, o CPF passou a ser a chave
/// natural civil da Pessoa (substituindo a matrícula, que migrou para o
/// exercício). Lixo neste campo corromperia a deduplicação de pessoas
/// — justamente o que viabiliza reaproveitar um cadastro existente na
/// admissão. Por isso validamos formato e dígitos verificadores, e
/// armazenamos SEMPRE normalizado (só dígitos), garantindo comparação
/// idêntica em SQLite e PostgreSQL.
/// </summary>
public static class Cpf
{
    /// <summary>Remove tudo que não é dígito (pontos, traços, espaços).</summary>
    public static string Normalizar(string? cpf) =>
        new((cpf ?? string.Empty).Where(char.IsDigit).ToArray());

    /// <summary>
    /// Valida um CPF pelos dois dígitos verificadores (algoritmo da
    /// Receita Federal). Rejeita tamanho diferente de 11 e as sequências
    /// repetidas (000..., 111...), que passam na conta mas são inválidas.
    /// </summary>
    public static bool EhValido(string? cpf)
    {
        var digitos = Normalizar(cpf);
        if (digitos.Length != 11) return false;
        if (digitos.All(c => c == digitos[0])) return false;

        static int CalcularDigito(string parcial, int pesoInicial)
        {
            var soma = 0;
            for (var i = 0; i < parcial.Length; i++)
                soma += (parcial[i] - '0') * (pesoInicial - i);
            var resto = soma % 11;
            return resto < 2 ? 0 : 11 - resto;
        }

        var primeiro = CalcularDigito(digitos[..9], 10);
        var segundo = CalcularDigito(digitos[..10], 11);
        return primeiro == digitos[9] - '0' && segundo == digitos[10] - '0';
    }
}
